### ISSUE TITLE
Create a Proxy for External Secrets

### ISSUE THREAD

**Issue by ianpurton** [purton-tech/cloak#90]
It would support push and pull via Rest.

We can then decrypt secrets.

**Comment by ianpurton** [comment]
@farazfazli Hi Faraz. If you wanted to you could take a look at this issue. It's at the system design stage at the moment so any feedback would be really appreciated.

**Comment by farazfazli** [comment]
Regarding possibility #1, does this mean that it would be calling cloak's API from within Kubernetes? Just trying to understand the flow better.

**Comment by ianpurton** [comment]
> Regarding possibility 1, does this mean that it would be calling cloak's API from within Kubernetes? Just trying to understand the flow better.

Yes. That API only partially exists at the moment. So it means extending the existing API at https://github.com/purton-tech/cloak/tree/main/crates/grpc-api

I think that needs to happen for both possibilities.

**Comment by farazfazli** [comment]
Hm okay, that makes sense. If you can wrap up the necessary APIs in Rust and provide some documentation, I can help out with writing the Go piece which calls the Cloak API.

**Comment by ianpurton** [comment]
@farazfazli So I've been thinking a bit more about this.

What I'm thinking now is to try and keep as much of the code in Rust as possible. I also realised half of the API is already there as the cloak cli tool calls the API to decrypt secrets.

So the architecture would look something like.

## a new crate

`crates/external-secrets-proxy`

This would be a small Axum app that initially has just one API end point i.e. `/secrets/read`

It would take a private key as an ENV var.
And the address of the server also as an ENV var. (defaulting to https://cloak.software)

It would then use the `crates/grpc-api` crate to make a gRPC call to the server to retrieve secretes.

It would return the secrets in a format that the external secrets operator webhook understands.

The cli is an example of how to call the API and decrypt secrets. https://github.com/purton-tech/cloak/blob/main/crates/cli/src/main.rs

## Testing

This will be tricky, but my idea so far is to try and incorporate https://kind.sigs.k8s.io/ into the CI/CD pipeline.
The somehow install the operator and tests.

## Release

So the main idea is that there should be some documentation added to external secrets operator to show people how to use cloak as a provider. They do this with markdown in this folder https://github.com/external-secrets/external-secrets/tree/main/docs/provider
